### ISSUE TITLE
Performance/config issue?

### ISSUE THREAD

**Issue by itzsimpl** [minio/minio#9825]
## Expected behavior

Minio Erasure Code in 12 drive setup (minioEC) performance to be more on par with Minio with zfs raidz2 pool of 12 drives (minioZ). Is there a way to help minioEC with cache drives? On upload there is a substantial difference also between minioZ and minioZC.

## Actual behaviour

Performance almost 10x slower on many small files upload.

## Steps to reproduce the behaviour

_small consists of 22500 wav.gz files with average length 40kB
_big is a 836M tar of _small 

upload:
time mc cp -r _small minioEC/test : 3.49 MiB/s 3m48s; real 3m48.402s
time mc cp -r _big minioEC/test : 185.19 MiB/s 4s; real 0m4.548s

time mc cp -r _small minioZ/test : 36.70 MiB/s 21s; real 0m21.770s
time mc cp -r _big minioZ/test : 422.67 MiB/s 1s; real 0m2.010s

time mc cp -r _small minioZC/test : 69.48 MiB/s 11s; real 0m11.518s
time mc cp -r _big minioZC/test : 541.42 MiB/s 1s; real 0m1.579s

download:
time mc cp -r minioEC/test/_small . : 32.63 MiB/s 24s; real 0m24.518s
time mc cp -r minioEC/test/_big . : 574.56 MiB/s 1s; real 0m1.491s

time mc cp -r minioZ/test/_small . : 56.41 MiB/s 14s; real 0m14.209s
time mc cp -r minioZ/test/_big . : 716.06 MiB/s 1s; real 0m1.202s

time mc cp -r minioZC/test/_small . : 57.03 MiB/s 13s; real 0m14.048s
time mc cp -r minioZC/test/_big . : 704.01 MiB/s 1s; real 0m1.223s

## minio --version
minio version RELEASE.2020-06-12T00-06-19Z

## mc --version
mc version RELEASE.2020-05-28T23-43-36Z

## System information

systemA:
12x 6TB SAS 12Gb/s
2x 512GB nvme 

minioEC:
mkfs.ext4 /dev/sda
...
mkfs.ext4 /dev/sdl
mount /dev/sda /mnt/minio_a
...
mount /dev/sdl /mnt/minio_l
minio --address :80 /mnt/minio_{a..l}

minioZ:
minio binary from 2020-06-12T00-06-19Z.
zpool create minio_z ridez2 /dev/sd{a..l}
minio --address :80 /minio_z

minioZC:
minio binary from 2020-06-12T00-06-19Z.
zpool create minio_zc ridez2 /dev/sd{a..l}
zpool add minio_zc log /dev/nvme0n1
zpool add minio_zc cache /dev/nvme1n1
zfs set compression=lz4 minio_zc
minio --address :80 /minio_zc

system B
10GbE connection to system A
both systems have lots of RAM and CPU power

**Comment by harshavardhana** [comment]
Don't use ZFS, just give all drives to MinIO on JBOD mode and use XFS (our recommended FS)

**Comment by itzsimpl** [comment]
> Don't use ZFS, just give all drives to MinIO on JBOD mode and use XFS (our recommended FS)

I had the same low performance for uploads even with XFS.

**Comment by harshavardhana** [comment]
To perform real numbers look at warp https://github.com/minio/warp

**Comment by itzsimpl** [comment]
Thank you for the link; I've ran the tests with warp mixed, and here are the results: 

minioZC:
Mixed operations.

Operation: DELETE
 * Operations: 3333 (10.1% of operations)
 * Throughput: 17.28 obj/s (3m12.954s, starting 16:38:41 UTC)

Operation: STAT
 * Operations: 9996 (30.4% of operations)
 * Throughput: 51.80 obj/s (3m13.004s, starting 16:38:41 UTC)

Operation: GET
 * Operations: 14956 (45.4% of operations)
 * Throughput: 776.56 MiB/s, 77.66 obj/s (3m13.025s, starting 16:38:41 UTC)

Operation: PUT
 * Operations: 5006 (15.2% of operations)
 * Throughput: 259.56 MiB/s, 25.96 obj/s (3m13.047s, starting 16:38:41 UTC)

Cluster Total:  1029.22 MiB/s, 171.76 obj/s (3m11.736s, starting 16:38:42 UTC)


minioZ:
Mixed operations.

Operation: STAT
 * Operations: 3927 (30.7% of operations)
 * Throughput: 67.80 obj/s (57.952s, starting 16:54:38 UTC)

Operation: DELETE
 * Operations: 1318 (10.3% of operations)
 * Throughput: 22.76 obj/s (58.002s, starting 16:54:38 UTC)

Operation: GET
 * Operations: 5873 (46.0% of operations)
 * Throughput: 1016.98 MiB/s, 101.70 obj/s (58.004s, starting 16:54:38 UTC)

Operation: PUT
 * Operations: 1960 (15.3% of operations)
 * Throughput: 339.26 MiB/s, 33.93 obj/s (57.94s, starting 16:54:38 UTC)

Cluster Total:  1349.25 MiB/s, 225.16 obj/s (56.75s, starting 16:54:39 UTC)


minioEC: [this time drives formatted XFS]
Mixed operations.

Operation: STAT
 * Operations: 5698 (31.2% of operations)
 * Throughput: 47.10 obj/s (2m1.012s, starting 17:04:29 UTC)

Operation: DELETE
 * Operations: 1898 (10.4% of operations)
 * Throughput: 15.70 obj/s (2m1.011s, starting 17:04:29 UTC)

Operation: GET
 * Operations: 8530 (46.6% of operations)
Errors: 1
 * Throughput: 705.67 MiB/s, 70.57 obj/s (2m0.94s, starting 17:04:29 UTC)

Operation: PUT
 * Operations: 2802 (15.3% of operations)
 * Throughput: 234.42 MiB/s, 23.44 obj/s (2m0.994s, starting 17:04:29 UTC)

Cluster Total:  940.69 MiB/s, 157.19 obj/s (1m56.34s, starting 17:04:32 UTC)

* repeated run, due to the error (connection reset by peer) that happened during the first run
Mixed operations.

Operation: STAT
 * Operations: 4786 (31.4% of operations)
 * Throughput: 46.55 obj/s (1m42.865s, starting 17:08:54 UTC)

Operation: GET
 * Operations: 7163 (47.0% of operations)
 * Throughput: 696.36 MiB/s, 69.64 obj/s (1m43.019s, starting 17:08:54 UTC)

Operation: DELETE
 * Operations: 1593 (10.5% of operations)
 * Throughput: 15.50 obj/s (1m42.898s, starting 17:08:54 UTC)

Operation: PUT
 * Operations: 2349 (15.4% of operations)
 * Throughput: 230.25 MiB/s, 23.02 obj/s (1m42.661s, starting 17:08:54 UTC)

Cluster Total:  928.36 MiB/s, 155.00 obj/s (1m38.356s, starting 17:08:57 UTC)

**Comment by harshavardhana** [comment]
NOTE: to get distributed numbers you need to run them with higher concurrency and also with equal number of client nodes.

Look at how to run distributed numbers using warp.

Increase the default size from 10MiB to 32MiB or 64MiB

**Comment by itzsimpl** [comment]
I apologise, but can you be more specific? I'm running warp via `docker run minio/warp`. Could you provide the parameters that should be run? 

I'm running one minio host (changing the config during runs) and one client (minio/warp).

**Comment by harshavardhana** [comment]
> I apologise, but can you be more specific? I'm running warp via `docker run minio/warp`. Could you provide the parameters that should be run?
> 
> I'm running one minio host (changing the config during runs) and one client (minio/warp).

Increase the file size for PUT and GET --obj.size 32MiB ?

**Comment by harshavardhana** [comment]
Also what you are seeing is the parity also being written basically 250MiB .. is basically 500MIB writes to disk with default parity.

This is perhaps shows how much is each drive capable really..  

So looks like it is a good number for the drives you have, you will get better numbers from more concurrency and bigger filesizes.

**Comment by itzsimpl** [comment]
Ok, I assume that by increasing the obj.size (leaving out obj.randsize) the test will be more skewed towards larger files, thus reducing the influence of transmitting small files (where there's really a big difference between in displayed performance of the underlying backend; xfs jbod vs zfs).

"Also what you are seeing is the parity also being written basically 250MiB .. is basically 500MIB writes to disk with default parity."
So basically what you're saying is that for this configuration the numbers that should be compared are throughput * 2 for minioEC vs. throughput * 1.2 for minioZ?

**Comment by itzsimpl** [comment]
I've been playing a little bit more; "Also what you are seeing is the parity also being written basically 250MiB .. is basically 500MIB writes to disk with default parity."
If I understand correctly by setting MINIO_STORAGE_CLASS_RRS=EC:2 and MINIO_STORAGE_CLASS_RRS=EC:2 the minioEC and minioZ should write the same amount of parity bits:

MinioEC:2
Mixed operations.

Operation: STAT
 * Operations: 6087 (30.9% of operations)
 * Throughput: 50.74 obj/s (2m0.009s, starting 20:11:10 UTC)

Operation: DELETE
 * Operations: 2030 (10.3% of operations)
 * Throughput: 16.95 obj/s (1m59.874s, starting 20:11:10 UTC)

Operation: GET
 * Operations: 9137 (46.4% of operations)
Errors: 1
 * Throughput: 762.01 MiB/s, 76.20 obj/s (2m0.021s, starting 20:11:10 UTC)

Operation: PUT
 * Operations: 3003 (15.3% of operations)
 * Throughput: 253.09 MiB/s, 25.31 obj/s (1m59.867s, starting 20:11:10 UTC)

Cluster Total:  1020.63 MiB/s, 170.65 obj/s (1m55.311s, starting 20:11:12 UTC)
  
which is still less then what minioZ achieves:
Mixed operations.                                                                                                                                                      
                                                                                                                                                                       
Operation: GET                                                                                                                                                         
 * Operations: 5666 (46.1% of operations)                                                                                                                              
 * Throughput: 1014.95 MiB/s, 101.49 obj/s (55.988s, starting 20:33:40 UTC)                                                                                            
                                                                                                                                                                       
Operation: STAT                                                                                                                                                        
 * Operations: 3801 (30.9% of operations)                                                                                                                              
 * Throughput: 68.01 obj/s (55.993s, starting 20:33:40 UTC)                                                                                                            
                                                                                                                                                                       
Operation: DELETE                                                                                                                                                      
 * Operations: 1267 (10.3% of operations)                                                                                                                              
 * Throughput: 22.69 obj/s (55.928s, starting 20:33:40 UTC)                                                                                                            
                                                                                                                                                                       
Operation: PUT                                                                                                                                                         
 * Operations: 1893 (15.4% of operations)                                                                                                                              
 * Throughput: 339.02 MiB/s, 33.90 obj/s (55.903s, starting 20:33:40 UTC)                                                                                              
                                                                                                                                                                       
Cluster Total:  1342.29 MiB/s, 223.92 obj/s (54.934s, starting 20:33:40 UTC) 


As suspected, setting obj.size=64MB does give better throughput to minioEC, and with this setting minioZ and minioEC achieve almost identical results:
Mixed operations.                                                                                                                                                      
                                                                                                                                                                       
Operation: STAT                                                                                                                                                        
 * Operations: 3657 (31.3% of operations)                                                                                                                              
 * Throughput: 12.29 obj/s (4m57.803s, starting 19:53:47 UTC)                                                                                                          
                                                                                                                                                                       
Operation: DELETE                                                                                                                                                      
 * Operations: 1226 (10.5% of operations)                                                                                                                              
 * Throughput: 4.13 obj/s (4m57.654s, starting 19:53:47 UTC)                                                                                                           
                                                                                                                                                                       
Operation: GET                                                                                                                                                         
 * Operations: 5449 (46.6% of operations)                                                                                                                              
 * Throughput: 1122.21 MiB/s, 18.39 obj/s (4m57.968s, starting 19:53:47 UTC)                                                                                           
                                                                                                                                                                       
Operation: PUT                                                                                                                                                         
 * Operations: 1833 (15.7% of operations)                                                                                                                              
 * Throughput: 376.50 MiB/s, 6.17 obj/s (4m57.923s, starting 19:53:47 UTC)                                                                                             
                                                                                                                                                                       
Cluster Total:  1481.82 MiB/s, 40.65 obj/s (4m47.416s, starting 19:53:53 UTC)

**Comment by harshavardhana** [comment]
> I've been playing a little bit more; "Also what you are seeing is the parity also being written basically 250MiB .. is basically 500MIB writes to disk with default parity."
> If I understand correctly by setting MINIO_STORAGE_CLASS_RRS=EC:2 and MINIO_STORAGE_CLASS_RRS=EC:2 the minioEC and minioZ should write the same amount of parity bits:
> 
> MinioEC:2
> Mixed operations.
> 
> Operation: STAT
> 
> * Operations: 6087 (30.9% of operations)
> * Throughput: 50.74 obj/s (2m0.009s, starting 20:11:10 UTC)
> 
> Operation: DELETE
> 
> * Operations: 2030 (10.3% of operations)
> * Throughput: 16.95 obj/s (1m59.874s, starting 20:11:10 UTC)
> 
> Operation: GET
> 
> * Operations: 9137 (46.4% of operations)
>   Errors: 1
> * Throughput: 762.01 MiB/s, 76.20 obj/s (2m0.021s, starting 20:11:10 UTC)
> 
> Operation: PUT
> 
> * Operations: 3003 (15.3% of operations)
> * Throughput: 253.09 MiB/s, 25.31 obj/s (1m59.867s, starting 20:11:10 UTC)
> 
> Cluster Total: 1020.63 MiB/s, 170.65 obj/s (1m55.311s, starting 20:11:12 UTC)
> 
> which is still less then what minioZ achieves:
> Mixed operations.
> 
> Operation: GET
> 
> * Operations: 5666 (46.1% of operations)
> * Throughput: 1014.95 MiB/s, 101.49 obj/s (55.988s, starting 20:33:40 UTC)
> 
> Operation: STAT
> 
> * Operations: 3801 (30.9% of operations)
> * Throughput: 68.01 obj/s (55.993s, starting 20:33:40 UTC)
> 
> Operation: DELETE
> 
> * Operations: 1267 (10.3% of operations)
> * Throughput: 22.69 obj/s (55.928s, starting 20:33:40 UTC)
> 
> Operation: PUT
> 
> * Operations: 1893 (15.4% of operations)
> * Throughput: 339.02 MiB/s, 33.90 obj/s (55.903s, starting 20:33:40 UTC)
> 
> Cluster Total: 1342.29 MiB/s, 223.92 obj/s (54.934s, starting 20:33:40 UTC)
> 
> As suspected, setting obj.size=64MB does give better throughput to minioEC, and with this setting minioZ and minioEC achieve almost identical results:
> Mixed operations.
> 
> Operation: STAT
> 
> * Operations: 3657 (31.3% of operations)
> * Throughput: 12.29 obj/s (4m57.803s, starting 19:53:47 UTC)
> 
> Operation: DELETE
> 
> * Operations: 1226 (10.5% of operations)
> * Throughput: 4.13 obj/s (4m57.654s, starting 19:53:47 UTC)
> 
> Operation: GET
> 
> * Operations: 5449 (46.6% of operations)
> * Throughput: 1122.21 MiB/s, 18.39 obj/s (4m57.968s, starting 19:53:47 UTC)
> 
> Operation: PUT
> 
> * Operations: 1833 (15.7% of operations)
> * Throughput: 376.50 MiB/s, 6.17 obj/s (4m57.923s, starting 19:53:47 UTC)
> 
> Cluster Total: 1481.82 MiB/s, 40.65 obj/s (4m47.416s, starting 19:53:53 UTC)

Not RRS it is _STANDARD Afaics there is no issue, it's what your drives are capable of. 

So that's basically 370*2 = 740 is your actual write throughput.

**Comment by itzsimpl** [comment]
> Not RRS it is _STANDARD...

I apologise, made a mistake in copy pasting, I've set both to EC:2. 

What I'm noticing is that at least on smaller files, there seems to be a noticeable difference in what type of setup MinIO Server runs on (Erasure Code or zfs), and it's visible both via minio/warp, as well as by a normal user via `mc cp`.

**Comment by harshavardhana** [comment]
> > Not RRS it is _STANDARD...
> 
> I apologise, made a mistake in copy pasting, I've set both to EC:2.
> 
> What I'm noticing is that at least on smaller files, there seems to be a noticeable difference in what type of setup MinIO Server runs on (Erasure Code or zfs), and it's visible both via minio/warp, as well as by a normal user via `mc cp`.

Small files you need to bring in some SSDs or NVMe . HDDs are not that suitable for small file workload.

**Comment by harshavardhana** [comment]
> > > Not RRS it is _STANDARD...
> > 
> > 
> > I apologise, made a mistake in copy pasting, I've set both to EC:2.
> > What I'm noticing is that at least on smaller files, there seems to be a noticeable difference in what type of setup MinIO Server runs on (Erasure Code or zfs), and it's visible both via minio/warp, as well as by a normal user via `mc cp`.
> 
> Small files you need to bring in some SSDs or NVMe . HDDs are not that suitable for small file workload.

I don't think we have any issue here, let's close this issue. It is working as expected for the hardware.

**Comment by itzsimpl** [comment]
That was my initial question; Is there a way to add SSD or NVMe cache to a MinIO Server running EC on pool of rust?

However, this still does not answer the source of the difference in performance between two, basically, equal setups.

**Comment by harshavardhana** [comment]
> However, this still does not answer the source of the difference in performance between two, basically, equal setups.

ZFS does its own stuff which changes performance behavior. 

> That was my initial question; Is there a way to add SSD or NVMe cache to a MinIO Server running EC on pool of rust?

There is no such thing possible adding caching is superfluous, and it is not useful the issue here is the hardware is not capable of what you want.

**Comment by itzsimpl** [comment]
Based on observations of how much the network was saturated during the minio/warp runs, and how much data the drives were actually writing, it seemed that in the case of zfs a lot more data was being written. Hence the higher numbers. It seems to me that the hw might be capable of more than what I'm getting.
